storage/rocksdb/ha_rocksdb.cc

```diff
@@ -9556,6 +9809,247 @@ int ha_rocksdb::index_read_intern(uchar *const buf, const uchar *const key,
+  rocksdb::Slice slice(slice_arg);
+  *use_locking_iterator = false;
+
+  if (!m_use_range_locking) {
+    return 0;
+  }
+  bool big_range = false;
+
+  /*
+    The 'slice' has the left endpoint of the range to lock.
+    Figure out the right endpoint.
+  */
+
+  if (find_flag == HA_READ_KEY_EXACT || find_flag == HA_READ_PREFIX_LAST) {
+    if (slice.size() == Rdb_key_def::INDEX_NUMBER_SIZE) {
+      // This is a full table/index scan
+      // (in case of HA_READ_PREFIX_LAST, a reverse-ordered one)
+      start_has_inf_suffix = false;
+      big_range = true;
+    } else {
+      /*
+        HA_READ_KEY_EXACT:
+          This is "key_part= const" interval. We need to lock this range:
+          (lookup_value, -inf) < key < (lookup_value, +inf)
+        HA_READ_PREFIX_LAST:
+          We get here for queries like:
+
+            select * from t1 where pk1=const order by pk1 desc for update
+
+          assuming this uses an index on (pk1, ...).
+          We get end_key=nullptr.
+      */
+      start_has_inf_suffix = false;
+      end_has_inf_suffix = true;
+      end_slice = slice;
+    }
+  } else if (find_flag == HA_READ_PREFIX_LAST_OR_PREV) {
+    /*
+       We get here for queries like:
+
+         select * from t1 where pk1=const1 and pk2 between const2 and const3
+         order by pk1 desc
+         for update
+
+       assuming this uses an index on (pk1, pk2).
+       The slice has the right endpoint: {const1, const3}
+       the end_key has the left endpoint: {const1, const2}.
+    */
+
+    // Move the right endpoint from slice to end_slice
+    end_slice = slice;
+
+    // Pack the left endpoint and make "slice" point to it
+    uchar pack_buffer[MAX_KEY_LENGTH];
+    uint end_slice_size = kd.pack_index_tuple(
+        table, pack_buffer, end_slice_buf, end_key->key, end_key->keypart_map);
+    slice =
+        rocksdb::Slice(reinterpret_cast<char *>(end_slice_buf), end_slice_size);
+    start_has_inf_suffix = false;
+    end_has_inf_suffix = true;
+  } else if (find_flag == HA_READ_BEFORE_KEY) {
+    /*
+      We get here for queries like
+        select * from t1
+        where              pk <1007 order by pk desc limit 2 for update
+        select * from t1
+        where pk >=800 and pk <1007 order by pk desc limit 2 for update
+    */
+
+    // Move the right endpoint from slice to end_slice
+    end_slice = slice;
+
+    if (end_key) {
+      uchar pack_buffer[MAX_KEY_LENGTH];
+      uint end_slice_size =
+          kd.pack_index_tuple(table, pack_buffer, end_slice_buf, end_key->key,
+                              end_key->keypart_map);
+
+      slice = rocksdb::Slice(reinterpret_cast<char *>(end_slice_buf),
+                             end_slice_size);
+
+      // end_has_inf_suffix is false, because we're looking key<const
+    } else {
+      uint end_slice_size;
+      kd.get_infimum_key(end_slice_buf, &end_slice_size);
+      slice = rocksdb::Slice((char *)end_slice_buf, end_slice_size);
```

Comment:
Is this work useful if `big_range == true`?

Also, I wonder if it makes sense to pull out the cases where big_range is true at the very beginning. It seems like `big_range` is true iff `end_key` is nullptr, and would reduce all these if checks here. There are some places where we set `big_range` for unknown range types, but arguably those should just error out.

Actually, I think `end_key` is also nullptr for the "prefix" flags, so it's not true that `big_range` iff `end_key == nullptr`. It may still be simpler to pull it out to the beginning though.